File: src/main/java/AoC2017_04.java
import java.util.Arrays;
import java.util.List;
import java.util.function.Predicate;

import com.github.pareronia.aoc.Counter;
import com.github.pareronia.aoc.Utils;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2017_04
            extends SolutionBase<List<String>, Integer, Integer> {

    private AoC2017_04(final boolean debug) {
        super(debug);
    }

    public static AoC2017_04 create() {
        return new AoC2017_04(false);
    }

    public static AoC2017_04 createDebug() {
        return new AoC2017_04(true);
    }
    
    @Override
    protected List<String> parseInput(final List<String> inputs) {
        return inputs;
    }

    private int solve(
            final List<String> input,
            final Predicate<List<String>> strategy
    ) {
        return (int) input.stream()
            .map(s -> Arrays.asList(s.split(" ")))
            .filter(strategy::test)
            .count();
    }

    @Override
    public Integer solvePart1(final List<String> input) {
        final Predicate<List<String>> hasNoDuplicateWords = words ->
            new Counter<>(words).values().stream().noneMatch(v -> v > 1L);
        return solve(input, hasNoDuplicateWords);
    }
    
    @Override
    public Integer solvePart2(final List<String> input) {
        final Predicate<List<String>> hasNoAnagrams = words ->
            words.size() == words.stream()
                .map(w -> new Counter<>(Utils.asCharacterStream(w)))
                .distinct().count();
        return solve(input, hasNoAnagrams);
    }

    @Samples({
        @Sample(method = "part1", input=TEST1, expected = "1"),
        @Sample(method = "part1", input=TEST2, expected = "0"),
        @Sample(method = "part1", input=TEST3, expected = "1"),
        @Sample(method = "part2", input=TEST4, expected = "1"),
        @Sample(method = "part2", input=TEST5, expected = "0"),
        @Sample(method = "part2", input=TEST6, expected = "1"),
        @Sample(method = "part2", input=TEST7, expected = "1"),
        @Sample(method = "part2", input=TEST8, expected = "0"),
    })
    public static void main(final String[] args) throws Exception {
        AoC2017_04.create().run();
    }
    
    private static final String TEST1 = "aa bb cc dd ee";
    private static final String TEST2 = "aa bb cc dd aa";
    private static final String TEST3 = "aa bb cc dd aaa";
    private static final String TEST4 = "abcde fghij";
    private static final String TEST5 = "abcde xyz ecdab";
    private static final String TEST6 = "a ab abc abd abf abj";
    private static final String TEST7 = "iiii oiii ooii oooi oooo";
    private static final String TEST8 = "oiii ioii iioi iiio";
}

File: src/main/cpp/2017/04/AoC2017_04.cpp
#include <bits/stdc++.h>
#include "../../aoc/aoc.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;

const vector<string> TEST1 = { "aa bb cc dd ee" };
const vector<string> TEST2 = { "aa bb cc dd aa" };
const vector<string> TEST3 = { "aa bb cc dd aaa" };
const vector<string> TEST4 = { "abcde fghij" };
const vector<string> TEST5 = { "abcde xyz ecdab" };
const vector<string> TEST6 = { "a ab abc abd abf abj" };
const vector<string> TEST7 = { "iiii oiii ooii oooi oooo" };
const vector<string> TEST8 = { "oiii ioii iioi iiio" };

vector<string> split(const string& s) {
    istringstream iss(s);
    return {istream_iterator<string>{iss}, istream_iterator<string>{}};
}

bool hasNoDuplicateWords(const string &s) {
    map<string, uint> m;
    for (const string &token : split(s)) {
        m[token]++;
    }
    return count_if(m.begin(), m.end(),
                    [](pair<string, int> pair) { return pair.second > 1; })
            == 0;
}

bool hasNoAnagrams(const string &s) {
    set<map<char, uint>> letterCounts;
    uint wordCount = 0;
    for (const string &token : split(s)) {
        wordCount++;
        map<char, uint> m;
        for (auto i = token.begin(); i != token.end(); ++i) {
            m[*i]++;
        }
        letterCounts.insert(m);
    }
    return letterCounts.size() == wordCount;
}

template<typename Strategy>
int solve(const Strategy &strategy, const vector<string> &input) {
    return count_if(input.begin(), input.end(), strategy);
}

int part1(const vector<string> &input) {
    return solve(hasNoDuplicateWords, input);
}

int part2(const vector<string> &input) {
    return solve(hasNoAnagrams, input);
}

void samples() {
    assert(part1(TEST1) == 1);
    assert(part1(TEST2) == 0);
    assert(part1(TEST3) == 1);
    assert(part2(TEST4) == 1);
    assert(part2(TEST5) == 0);
    assert(part2(TEST6) == 1);
    assert(part2(TEST7) == 1);
    assert(part2(TEST8) == 0);
}

MAIN(2017, 4)
